pow/src/main/java/tech/pegasys/artemis/pow/DepositProcessingController.java

```diff
@@ -0,0 +1,210 @@
+                        + "have access to new deposits after genesis.",
+                    error));
+  }
+
+  private synchronized void onNewCanonicalBlockNumber(BigInteger latestCanonicalBlockNumber) {
+    if (latestCanonicalBlockNumber.compareTo(this.latestCanonicalBlockNumber) <= 0) {
+      return;
+    }
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    fetchLatestSubscriptionDeposits();
+  }
+
+  private synchronized void fetchLatestSubscriptionDeposits() {
+    if (isBlockByBlockModeOn) {
+      fetchLatestDepositsOneBlockAtATime();
+    } else fetchLatestSubscriptionDepositsOverRange();
+  }
+
+  private synchronized void fetchLatestSubscriptionDepositsOverRange() {
+    final BigInteger toBlock;
+    final BigInteger fromBlock;
+
+    if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+      return;
+    }
+    active = true;
+
+    fromBlock = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    toBlock = latestCanonicalBlockNumber;
+
+    depositFetcher
+        .fetchDepositsInRange(fromBlock, toBlock)
+        .finish(
+            __ -> onSubscriptionDepositRequestSuccessful(toBlock),
+            (err) -> onSubscriptionDepositRequestFailed(err, fromBlock, toBlock));
+  }
+
+  private synchronized void fetchLatestDepositsOneBlockAtATime() {
+    final BigInteger nextBlockNumber;
+
+    synchronized (DepositProcessingController.this) {
+      if (active || latestCanonicalBlockNumber.equals(latestSuccessfullyQueriedBlock)) {
+        return;
+      }
+      active = true;
+
+      nextBlockNumber = latestSuccessfullyQueriedBlock.add(BigInteger.ONE);
+    }
+
+    depositFetcher
+        .fetchDepositsInRange(nextBlockNumber, nextBlockNumber)
+        .thenCompose(
+            __ ->
+                eth1Provider.getGuaranteedEth1BlockFuture(
+                    UnsignedLong.valueOf(nextBlockNumber), asyncRunner))
+        .thenAccept(
+            block -> {
+              final BigInteger blockNumber = block.getNumber();
+              LOG.trace("Successfully fetched block {} for min genesis checking", blockNumber);
+              LOG.trace(
+                  "Seconds until min genesis block {}",
+                  Constants.MIN_GENESIS_TIME.minus(
+                      calculateCandidateGenesisTimestamp(block.getTimestamp())));
+              if (MinimumGenesisTimeBlockFinder.compareBlockTimestampToMinGenesisTime(block) >= 0) {
+                notifyMinGenesisTimeBlockReached(eth1EventsChannel, block);
+                isBlockByBlockModeOn = false;
+                LOG.debug("Switching back to fetching deposits by range");
+              }
+            })
+        .finish(
+            __ -> onSubscriptionDepositRequestSuccessful(nextBlockNumber),
+            (err) -> onSubscriptionDepositRequestFailed(err, nextBlockNumber));
+  }
+
+  private synchronized void onSubscriptionDepositRequestSuccessful(BigInteger requestToBlock) {
```

Comment:
nit: We're not using a subscription to get deposits anymore so these onSubscription* methods should just be onDepositRequestSuccessful etc.

Reply:
The request methods still get run after a block received from the subscription actually. So I'm on the side of keeping them.